registry/proxy/proxyblobstore.go

```diff
@@ -42,140 +37,107 @@ func setResponseHeaders(w http.ResponseWriter, length int64, mediaType string, d
+		return true, pbs.localStore.ServeBlob(ctx, w, r, dgst)
+	}
 
-			proxyMetrics.BlobPull(uint64(desc.Size))
-		}()
-		err := streamToClient(ctx, w, desc, bw)
-		if err != nil {
-			return err
-		}
+	return false, nil
 
-		proxyMetrics.BlobPush(uint64(desc.Size))
-		pbs.scheduler.AddBlob(dgst.String(), blobTTL)
-		return nil
-	}
+}
 
-	err = streamToClient(ctx, w, desc, bw)
+func (pbs *proxyBlobStore) ServeBlob(ctx context.Context, w http.ResponseWriter, r *http.Request, dgst digest.Digest) error {
+	served, err := pbs.serveLocal(ctx, w, r, dgst)
 	if err != nil {
+		context.GetLogger(ctx).Errorf("Error serving blob from local storage: %s", err.Error())
 		return err
 	}
-	proxyMetrics.BlobPush(uint64(desc.Size))
-	return nil
-}
-
-type cleanupFunc func()
-
-// getOrCreateBlobWriter will track which blobs are currently being downloaded and enable client requesting
-// the same blob concurrently to read from the existing stream.
-func getOrCreateBlobWriter(ctx context.Context, blobs distribution.BlobService, desc distribution.Descriptor) (distribution.BlobWriter, bool, cleanupFunc, error) {
-	mu.Lock()
-	defer mu.Unlock()
-	dgst := desc.Digest
-
-	cleanup := func() {
-		mu.Lock()
-		defer mu.Unlock()
-		inflight[dgst].refCount--
-
-		if inflight[dgst].refCount == 0 {
-			defer delete(inflight, dgst)
-			_, err := inflight[dgst].bw.Commit(ctx, desc)
-			if err != nil {
-				// There is a narrow race here where Commit can be called while this blob's TTL is expiring
-				// and its being removed from storage.  In that case, the client stream will continue
-				// uninterruped and the blob will be pulled through on the next request, so just log it
-				context.GetLogger(ctx).Errorf("Error committing blob: %q", err)
-			}
 
-		}
+	if served {
+		return nil
 	}
 
-	var bw distribution.BlobWriter
+	mu.Lock()
 	_, ok := inflight[dgst]
 	if ok {
-		bw = inflight[dgst].bw
-		inflight[dgst].refCount++
-		return bw, false, cleanup, nil
+		mu.Unlock()
+		_, err := pbs.copyContent(ctx, dgst, w)
+		return err
 	}
+	inflight[dgst] = struct{}{}
+	mu.Unlock()
+
+	var commitErr error
+	go func() {
+		defer func() {
+			mu.Lock()
+			delete(inflight, dgst)
+			mu.Unlock()
+		}()
 
-	var err error
-	bw, err = blobs.Create(ctx)
-	if err != nil {
-		return nil, false, nil, err
-	}
+		bw, err := pbs.localStore.Create(ctx)
+		if err != nil {
+			commitErr = err
+			return
+		}
+		desc, err := pbs.copyContent(ctx, dgst, bw)
+		if err != nil {
+			commitErr = err
+			return
+		}
 
-	inflight[dgst] = &inflightBlob{refCount: 1, bw: bw}
-	return bw, true, cleanup, nil
-}
+		_, err = bw.Commit(ctx, desc)
+		if err != nil {
+			commitErr = err
+		}
+		pbs.scheduler.AddBlob(dgst.String(), repositoryTTL)
+	}()
 
-func streamToStorage(ctx context.Context, remoteReader distribution.ReadSeekCloser, desc distribution.Descriptor, bw distribution.BlobWriter) error {
-	_, err := io.CopyN(bw, remoteReader, desc.Size)
-	if err != nil {
-		return err
+	if commitErr != nil {
```

Comment:
I'm confused by this code flow.

What we have is basically the following:

```go
var commitErr error
go func() {
    // ...
    commitErr = foo;
    // ...
}()
if commitErr != nil {
    // ...
}
```

Something seems wrong.